src/rtpghi~.c

```diff
@@ -12,147 +12,187 @@
-    post("length of c[]: %d", (M) * sizeof *(x->c));
-    post("s_n: %d", sp[0]->s_n);
-    
-    t_sample **dummy=x->out;
-    *dummy++=sp[1]->s_vec;
-    *dummy=sp[2]->s_vec;
-    
-    dsp_add(rtpghi_tilde_perform, 3,x, sp[0]->s_vec, sp[0]->s_n);
-}
+	ltfat_complex_s *c = (ltfat_complex_s *) x->c;
 
-void *rtpghi_tilde_new(t_symbol *s, int argc, t_atom *argv)
-{
+	int e = 0;
 
-  t_rtpghi_tilde *x = (t_rtpghi_tilde *)pd_new(rtpghi_tilde_class);
+	if ((s==NULL) || ((x->sta_pd)==NULL)) {
+	   pd_error(x, "arrays not initialised");
+	}
+	else
+	{
+		e = phaseret_rtpghi_execute_s(x->sta_pd, s, c);
+		
+		t_sample *out=(t_sample *) (w[3]);
+		t_sample *out1= (t_sample *) (w[4]);
 
-    x->W_pd=atom_getfloat(argv);
-    x->ol_pd=atom_getfloat(argv+1);
-    x->M_pd=atom_getfloat(argv+2);
-    x->tol_pd=atom_getfloat(argv+3);
-    x->do_causal_pd=atom_getfloat(argv+4);
-    x->window_type= *atom_getsymbol(argv+5);
-    x->sta_pd=NULL;
-    x-> c = NULL;
-    ltfat_int M = x->M_pd;
+		if (e == 0) {
+			while (n--) {
+				
+				if (n >= n/2-1){
+					*out++ = creal(*c);
+					*out1++ = cimag(*c++);
+				}
+				else {
+					
+					*out++ = 0.f;
+					*out1++ = 0.f;
+				}
+			}
+		}
+		else {
+			pd_error(x, "error of type: %d", e);
+		}
+	}
+
+ return (w+6);
+}
+
+void rtpghi_tilde_dsp(t_rtpghi_tilde *x, t_signal **sp)
+{
+	post("start DSP");
+    ltfat_int M = (ltfat_int) sp[0]->s_n;
     
     post("stft_length [M]: %d\n", M);
-    const char* win=x->window_type.s_name;
-    
-    LTFAT_FIRWIN window = ltfat_str2firwin(win);
+	const char* win = x->window_type_pd->s_name;
+	const char win_[4] = "hann";
+	LTFAT_FIRWIN window;
+	
+	if ((x->win_def) == 1) {
+		window = ltfat_str2firwin(win);
+	}
+	else {
+		window = ltfat_str2firwin(win_);
+	}
+	
     double gamma = phaseret_firwin2gamma(window, M);
     post("window normalization parameter [gamma]: %f\n", gamma);
     
-    ltfat_int W = x->W_pd;
-    post("number of audio channels [W]: %d\n", W);
-    
-    
     ltfat_int a = M/(x->ol_pd);
     post("hopsize: %d\n", a);
     
     double tol = x->tol_pd;
     post("Tolerance [tol]: %f\n", tol);
     
-    int do_causal = 0;
+    int do_causal = x->do_causal_pd;
     post("Causal yes=1, no=0, [do_causal]: %d\n", do_causal);
+	
+	
+	// Check for existing ltfat_state (to do: implement init only when input param.)
+	int init_s;
+	ltfat_int w = 1;
+	
+	if (x->sta_pd == NULL) {
+		init_s = phaseret_rtpghi_init_s(gamma, w, a , M, tol, do_causal, &(x->sta_pd));
+		post("failed to init, status %d", init_s);
+		if (init_s == 0) {
+			post("initialised rtpghi plan at adress: %p\n", &(x->sta_pd));
+		}
+		else {
+			post("failed to init, status %d", init_s);
+			pd_error(x, "failed to init, status %d", init_s);
+		}
+	}
+	else {
+		
+		post("destroyed state at adrr: %p\n", &(x->sta_pd));
+		phaseret_rtpghi_done_s(&(x->sta_pd));
+		init_s = phaseret_rtpghi_init_s(gamma, w, a , M, tol, do_causal, &(x->sta_pd));
+		if (init_s == 0) {
```

Comment:
it is of course possible to do it like this, but it's not very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), as you have to call the init func twice and make sure that you do it correctly all the time.

what you are doing right now is:
```C
if(x->resource) {
  free_resource(x->resource);
   x->resource = allocate_resource();
   check_resource(x->resource);
} else {
   x->resource = allocate_resource();
   check_resource(x->resource);
}
```

whereas a simplified version (that does the very same thing without duplicating code) looks like this:

if(x->resource) {
  free_resource(x->resource);
}
x->resource = allocate_resource();
check_resource(x->resource);
```

in practice, this would look more like this:

```C
if(changed) {
   free_resource(x->resource);
   x->resource = NULL;
}
if(!x->resource) {
   x->resource = allocate_resource();
   check_resource(x->resource);
}
```